[KADENCE USES IT] Bugfix: multisite token logic

defunctl opened the pull request:
## Main Changes

This PR completely changes the way Uplink treats multisite installations. See [the new multisite licensing documentation](https://github.com/stellarwp/uplink/blob/f334ac154c1f16cf2b8e768d06510c2263488921/docs/features.md#multisite-licenses).

- All token storing/disconnecting logic now requires a Resource slug in order to test whether the plugin is network activated or not, so we know where to store the token.
- Lots of new helper functions to provide a common interface for accessing Uplink functionality.
- 

## Other

- Fixes failing Replacement_Key tests due to renaming of base API url.

defunctl committed:
Fix HTTP mocks not working after base url was renamed.

defunctl committed:
Clean up test

defunctl committed:
Add ability to mock a plugin is activated

defunctl committed:
WIP: refactor token auth for multisite using plugin

defunctl committed:
Include slug in token disconnect logic

defunctl committed:
Refactor Authorized

defunctl committed:
WIP: conditional token storage

defunctl committed:
WIP: multisite logic refactor

defunctl committed:
Add get_license_key function

defunctl committed:
Reset multisite config to original state after each test

defunctl committed:
Refactor get_domain() to take into consideration unique subsites when in subfolder mode

defunctl committed:
Remove token check, replace with main site check

defunctl committed:
Add Config::allows_network_licenses() and update Multisite_Main_Site.php

defunctl committed:
Add a hash suffix to the main site URL when on multisite with network licenses

defunctl committed:
Delete old pipeline steps

defunctl committed:
Add more helper functions

defunctl committed:
Rename get_domain function

defunctl committed:
Move disconnect URL into its own method, add helper function

defunctl committed:
Allow getting the original domain name

defunctl committed:
Get original domain for download args

defunctl committed:
Remove get_token, update get_authorization_token

defunctl committed:
Delete unneeded Authorized class

defunctl committed:
Fix license manager caching

defunctl committed:
Clean up functions, add is_user_authorized()

defunctl committed:
Add and adjust functions, first pass at multisite licensing documentation

defunctl committed:
Rollback is_authorized() change, add a new function instead

defunctl committed:
Fix warning

defunctl committed:
Refactor allows_multisite_license() to accept an already made resource

defunctl committed:
Fix comment

defunctl committed:
Add set_license_key example

defunctl committed:
Fix set_license_key function

defunctl committed:
Remove stray ampersand

defunctl committed:
formatting

defunctl committed:
Attempt to allow network validation

defunctl committed:
Show the correct message based on network validation

defunctl committed:
submit to network settings if on the network_admin

defunctl committed:
Add is_license_valid() function

defunctl committed:
rename to validate_license, return entire response

defunctl committed:
Don't return null on empty license, validate allows a null key

defunctl committed:
Remove unused container property, add types

defunctl committed:
Add nullable type

defunctl committed:
Use dependency injection for Ajax

defunctl committed:
ensure we are multisite aware when fetching the license key for the settings field

defunctl committed:
Fix bad return type

defunctl committed:
Show default licensing messages if coming from the network admin

defunctl committed:
Store key status in network if required

defunctl committed:
Update the license key valid cache after updating key locally

defunctl committed:
Ensure key is network at network/local level when validating

defunctl committed:
Don't register settings during ajax requests

defunctl committed:
Ensure to validate license key before storing when callback connecting

defunctl committed:
Ensure key being checked is network aware during license validation

defunctl committed:
set_license_key() will properly update the validation state of the license

borkweb commented on src/Uplink/Auth/Authorizer.php:
```diff
@@ -2,35 +2,25 @@
 	public function can_auth(): bool {
-		return $this->pipeline->send( true )->thenReturn();
+		return (bool) apply_filters(
+			'stellarwp/uplink/' . Config::get_hook_prefix() . '/auth/user_check',
```
```suggestion
			'stellarwp/uplink/' . Config::get_hook_prefix() . '/auth/can_auth',
```

borkweb commented on src/Uplink/Auth/License/Pipeline/Traits/Multisite_Trait.php:
```diff
@@ -0,0 +1,83 @@
+use StellarWP\Uplink\Utils\Checks;
+
+trait Multisite_Trait {
+
+	private function get_main_site_url(): string {
+		return get_site_url( get_main_site_id() );
+	}
+
+	private function get_current_site_url(): string {
```
Docblocks

borkweb commented on src/Uplink/Config.php:
```diff
@@ -31,6 +31,27 @@ class Config {
+	 *
+	 * @var bool
+	 */
+	protected static $network_subfolder_license = false;
```
The `::reset()` method needs to reset all of the static var values back to defaults. (just commenting here, though the comment applies to all of the static vars.)

borkweb commented on src/admin-views/fields/settings.php:
```diff
@@ -25,7 +26,7 @@
 	<div class="stellarwp-uplink__settings">
 		<?php do_action( 'stellarwp/uplink/' . Config::get_hook_prefix(). '/license_field_before_form', $plugin->get_slug() ) ?>
-		<form method="post" action="options.php">
+		<form method="post" action="<?php echo esc_url( $form_action ) ?>">
```
```suggestion
		<form method="post" action="<?php echo esc_url( $form_action ); ?>">
```

borkweb reviewed (changes requested):
I know you aren't done here, but I took a glance and found a couple of small things.

defunctl committed:
Merge branch 'feature/authorization-caching' into bugfix/multisite-token-logic

# Conflicts:
#	src/Uplink/Config.php
#	src/Uplink/functions.php
#	tests/wpunit/ConfigTest.php

defunctl committed:
Update base testcase to fix merge issue and use config::reset()

defunctl committed:
Merge branch 'fix/ESM-70-fatal-error-on-ticket-transfer' into bugfix/multisite-token-logic

defunctl committed:
Merge branch 'main' into bugfix/multisite-token-logic

Commit:
Merge branch 'feat/enable-auto-updates' into bugfix/multisite-token-logic

defunctl committed:
Use unique actions for each resource slug for token connect/disconnect

defunctl committed:
Remove prefix from token connect/disconnect actions, we don't have this data in uplink origin

defunctl committed:
Bugfix: Ensure download_url property is not null before attempting to use a condition on it

defunctl committed:
Add action to disconnect controller

defunctl committed:
Properly fix Undefined property: stdClass::$download_url

defunctl committed:
Merge branch 'bugfix/multisite-token-logic' into bugfix/token-connect-disconnect-logic

defunctl committed:
Refactor to use uplink_slug and no longer use WordPress's action query var

defunctl committed:
Formatting

defunctl committed:
Use a constant for the auth action

defunctl committed:
Add ActionManagerTest.php

defunctl committed:
Move assertion into single loop

defunctl committed:
Use standard action format with prefix

defunctl committed:
Update connect/disconnect hook name

defunctl committed:
Fix action docblocks

defunctl committed:
call correct action in tests

defunctl committed:
Formatting

defunctl committed:
Fix spelling

defunctl committed:
Formatting

defunctl committed:
Merge pull request #59 from stellarwp/bugfix/token-connect-disconnect-logic

Bugfix: Use unique actions for each resource slug for token connect/disconnect

defunctl committed:
Merge remote-tracking branch 'origin/bugfix/prevent-empty-class' into bugfix/multisite-token-logic

defunctl committed:
Merge remote-tracking branch 'origin/feat/allow-disabling-updates' into bugfix/multisite-token-logic

defunctl committed:
Make authorized checks require a plugin slug

defunctl committed:
Fix missing variable, update README.md

defunctl committed:
Don't use slug for transient cache key

defunctl committed:
Only use the token to make the cache key

defunctl committed:
Refactor token disconnect logic to also delete the authorization cache

defunctl committed:
Properly build a cache hash with just a token, don't use prefix in query var

defunctl committed:
Merge pull request #64 from stellarwp/feature/ms/authorized-check-to-use-plugin-slug

Make authorized checks require a plugin slug

defunctl committed:
Merge branch 'main' into bugfix/multisite-token-logic

defunctl committed:
Show WP info

defunctl committed:
use the correct env var to call slic

defunctl committed:
Fix tests failing due to Unexpected incorrect usage notice for wp_add_privacy_policy_content.

defunctl committed:
Bump actions/checkout versions, fix composer cache dir

defunctl committed:
Manually bring in updated tests matrix

defunctl committed:
Fix type error in filter_upgrader_pre_download

defunctl committed:
Merge branch 'main' into bugfix/KAD-2308/updater-type-error

defunctl committed:
Merge branch 'bugfix/KAD-2308/updater-type-error' into bugfix/multisite-token-logic

# Conflicts:
#	.github/workflows/tests-php.yml

defunctl committed:
Fix typo

defunctl committed:
Merge branch 'bugfix/KAD-2308/updater-type-error' into bugfix/multisite-token-logic

borkweb commented on src/Uplink/Admin/Plugins_Page.php:
```diff
@@ -210,6 +210,9 @@ public function inject_info( $result, ?string $action = null, $args = null ) {
 		if ( ! $relevant ) {
 			return $result;
 		}
+		if ( apply_filters( 'stellarwp/uplink/' . $args->slug . '/prevent_update_check', false ) ) {
```
Just a line break!

```suggestion

		if ( apply_filters( 'stellarwp/uplink/' . $args->slug . '/prevent_update_check', false ) ) {
```

defunctl committed:
Force type for delete_transient kadence zendesk ticket 610295

defunctl committed:
Force set_transient type for good measure

defunctl committed:
Allow some HTML tags in notices, show error if nonce failed and litespeed plugin is active

defunctl committed:
Fix allowed tags

Commit:
add filter

Commit:
fix test

Commit:
updates with more escaping and not showing a button unless it's TEC or there is a link set

Commit:
test update

defunctl committed:
Use require_once as a statement

defunctl committed:
Remove deprecated config, ignore require_once errors

defunctl committed:
Bump action versions

defunctl committed:
Merge pull request #90 from stellarwp/chore/phpstan-updates

Chore: phpstan updates

defunctl committed:
Merge branch 'bugfix/multisite-token-logic' into bugfix/missing-filter

Commit:
Fix array

Commit:
Add test

Commit:
fix string

Commit:
target..

Commit:
Merge pull request #89 from stellarwp/bugfix/missing-filter

Add expired message filter

Commit:
Account for an empty siteurl in the database

Commit:
apply change

Commit:
Merge pull request #93 from stellarwp/bugfix/empty-domain

Bugfix/empty domain

Commit:
fix so filter works

Commit:
update the get_license_key function to be able to get from a file

Commit:
Fix in the license field

Commit:
Merge pull request #97 from stellarwp/fix/get-key-from-file

update the get_license_key function to be able to get from a file

Commit:
cache is auth

Commit:
It's possible token is an empty string so lets add each to this.

Commit:
fix test

Commit:
use error

Commit:
remove error response that won't work

defunctl committed:
Merge branch 'bugfix/multisite-token-logic' into feat/cache-is-auth

Commit:
Update tests/wpunit/API/V3/AuthorizerCacheTest.php

Co-authored-by: Justin <defunctl@users.noreply.github.com>

Commit:
Update tests/wpunit/API/V3/AuthorizerCacheTest.php

Co-authored-by: Justin <defunctl@users.noreply.github.com>

Commit:
Justin suggestion

Commit:
Merge pull request #98 from stellarwp/feat/cache-is-auth

Add caching to a false response for is auth to prevent requests every page load.

Commit:
fix for cache miss

get_transient doesn't return an integer, so we need to store and test against a string

defunctl committed:
Update to use constants

Commit:
Merge pull request #99 from stellarwp/bug/fix-auth-cache

fix for cache miss

Commit:
allow overriding info specific to wp format

Commit:
update for php

borkweb committed:
Merge pull request #95 from stellarwp/bugfix/api_base_url_filter

Bugfix/api base url filter

dave-green-uk commented:
@defunctl do you know what (if anything) is blocking this being reviewed and merged?

We could use it for: https://github.com/iconicwp/iconic-core/pull/50#issuecomment-3670503301

johnhooks commented:
@defunctl if you can resolve the conflicts, it would be very helpful. Thanks.
